accord-core/src/main/java/accord/impl/progresslog/WaitingState.java

```diff
@@ -0,0 +1,624 @@
+        return route.slice(ranges);
+    }
+
+    private static Route<?> awaitRoute(Route<?> slicedRoute, BlockedUntil blockedUntil)
+    {
+        return blockedUntil.waitsOn == HOME ? slicedRoute.homeKeyOnlyRoute() : slicedRoute;
+    }
+
+    private static Route<?> fetchRoute(Route<?> slicedRoute, Route<?> awaitRoute, BlockedUntil blockedUntil)
+    {
+        return blockedUntil.waitsOn == blockedUntil.fetchFrom ? awaitRoute : slicedRoute;
+    }
+
+    void setWaitingDone(DefaultProgressLog owner)
+    {
+        set(null, owner, CanApply, NoneExpected);
+        owner.clearActive(Waiting, txnId);
+        clearRetryCounter();
+    }
+
+    void setBlockedUntil(SafeCommandStore safeStore, DefaultProgressLog owner, BlockedUntil blockedUntil)
+    {
+        BlockedUntil currentlyBlockedUntil = blockedUntil();
+        if (blockedUntil.compareTo(currentlyBlockedUntil) > 0 || isUninitialised())
+        {
+            clearAwaitState();
+            clearRetryCounter();
+            owner.clearActive(Waiting, txnId);
+            set(safeStore, owner, blockedUntil, Queued);
+        }
+    }
+
+    void record(DefaultProgressLog owner, SaveStatus newSaveStatus)
+    {
+        BlockedUntil currentlyBlockedUntil = blockedUntil();
+        if (currentlyBlockedUntil.minSaveStatus.compareTo(newSaveStatus) <= 0)
+        {
+            boolean isDone = newSaveStatus.hasBeen(Status.PreApplied);
+            set(null, owner, isDone ? CanApply : currentlyBlockedUntil, NoneExpected);
+            if (isDone)
+                maybeRemove(owner);
+            owner.clearActive(Waiting, txnId);
+        }
+    }
+
+    final void runWaiting(SafeCommandStore safeStore, SafeCommand safeCommand, DefaultProgressLog owner)
+    {
+        run(safeStore, safeCommand, owner);
+    }
+
+    private void run(SafeCommandStore safeStore, SafeCommand safeCommand, DefaultProgressLog owner)
+    {
+        BlockedUntil blockedUntil = blockedUntil();
+        Command command = safeCommand.current();
+        Invariants.checkState(!owner.hasActive(Waiting, txnId));
+        Invariants.checkState(command.saveStatus().compareTo(blockedUntil.minSaveStatus) < 0, "Command has met desired criteria but progress log entry has not been cancelled");
+
+        set(safeStore, owner, blockedUntil, Querying);
+        TxnId txnId = safeCommand.txnId();
+        // first make sure we have enough information to obtain the command locally
+        Timestamp executeAt = command.executeAtIfKnown();
+        EpochSupplier toLocalEpoch = toLocalEpoch(safeStore, txnId, blockedUntil, command, executeAt);
+        Participants<?> fetchKeys = Invariants.nonNull(command.maxContactable());
+
+        if (!Route.isRoute(fetchKeys))
+        {
+            fetchRoute(owner, blockedUntil, txnId, executeAt, toLocalEpoch, fetchKeys);
+            return;
+        }
+
+        Route<?> route = Route.castToRoute(fetchKeys);
+        if (homeSatisfies().compareTo(blockedUntil) < 0)
```

Comment:
We try to avoid lots of wasted distributed work by first waiting until we at least know of some home shard that has the ability to answer our query. This is stated briefly at the class doc, but I agree it is a bit lacking. I will add a comment on the method.